From c98076edd90596c3747d4c8ecad60876007f931d Mon Sep 17 00:00:00 2001
From: Igor Opaniuk <igor.opaniuk@foundries.io>
Date: Thu, 12 Aug 2021 16:14:46 +0300
Subject: [PATCH 167/322] [FIO toup] apalis-imx8: add support of SPL build

Sync with NXP reference board iMX8QM-MEK, where SPL build is
supported.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
---
 board/toradex/apalis-imx8/Makefile            |  1 +
 .../apalis-imx8/apalis-imx8-imximage.cfg      |  4 +-
 board/toradex/apalis-imx8/apalis-imx8.c       | 25 ++++++++
 board/toradex/apalis-imx8/spl.c               | 63 +++++++++++++++++++
 configs/apalis-imx8_defconfig                 |  1 +
 include/configs/apalis-imx8.h                 | 34 +++++++++-
 6 files changed, 122 insertions(+), 6 deletions(-)
 create mode 100644 board/toradex/apalis-imx8/spl.c

diff --git a/board/toradex/apalis-imx8/Makefile b/board/toradex/apalis-imx8/Makefile
index a8c3eb7240b..18dcbd2d272 100644
--- a/board/toradex/apalis-imx8/Makefile
+++ b/board/toradex/apalis-imx8/Makefile
@@ -4,3 +4,4 @@
 #
 
 obj-y += apalis-imx8.o
+obj-$(CONFIG_SPL_BUILD) += spl.o
diff --git a/board/toradex/apalis-imx8/apalis-imx8-imximage.cfg b/board/toradex/apalis-imx8/apalis-imx8-imximage.cfg
index b8f0f3dc10f..81b870af765 100644
--- a/board/toradex/apalis-imx8/apalis-imx8-imximage.cfg
+++ b/board/toradex/apalis-imx8/apalis-imx8-imximage.cfg
@@ -19,6 +19,4 @@ CONTAINER
 /* Add scfw image with exec attribute */
 IMAGE SCU mx8qm-apalis-scfw-tcm.bin
 /* Add ATF image with exec attribute */
-IMAGE A35 bl31.bin 0x80000000
-/* Add U-Boot image with load attribute */
-DATA A35 u-boot-dtb.bin 0x80020000
+IMAGE A35 spl/u-boot-spl.bin 0x00100000
diff --git a/board/toradex/apalis-imx8/apalis-imx8.c b/board/toradex/apalis-imx8/apalis-imx8.c
index f2204a72e6c..6dca728f952 100644
--- a/board/toradex/apalis-imx8/apalis-imx8.c
+++ b/board/toradex/apalis-imx8/apalis-imx8.c
@@ -3,6 +3,7 @@
  * Copyright 2019 Toradex
  */
 
+#include <command.h>
 #include <common.h>
 #include <cpu_func.h>
 #include <init.h>
@@ -41,6 +42,7 @@ DECLARE_GLOBAL_DATA_PTR;
 #define GPIO_PAD_CTRL	((SC_PAD_CONFIG_NORMAL << PADRING_CONFIG_SHIFT) | (SC_PAD_ISO_OFF << PADRING_LPCONFIG_SHIFT) \
 						| (SC_PAD_28FDSOI_DSE_DV_HIGH << PADRING_DSE_SHIFT) | (SC_PAD_28FDSOI_PS_PU << PADRING_PULL_SHIFT))
 
+#ifndef CONFIG_SPL_BUILD
 #define PCB_VERS_DEFAULT	((SC_PAD_CONFIG_NORMAL << PADRING_CONFIG_SHIFT) | \
 				 (SC_PAD_ISO_OFF << PADRING_LPCONFIG_SHIFT) | \
 				 (SC_PAD_28FDSOI_PS_PD << PADRING_PULL_SHIFT) | \
@@ -60,6 +62,7 @@ static iomux_cfg_t pcb_vers_default[] = {
 	SC_P_MIPI_DSI0_GPIO0_00 | MUX_MODE_ALT(3) | MUX_PAD_CTRL(PCB_VERS_DEFAULT),
 	SC_P_MIPI_DSI0_GPIO0_01 | MUX_MODE_ALT(3) | MUX_PAD_CTRL(PCB_VERS_DEFAULT),
 };
+#endif
 
 static iomux_cfg_t uart1_pads[] = {
 	SC_P_UART1_RX | MUX_PAD_CTRL(UART_PAD_CTRL),
@@ -225,6 +228,7 @@ void board_late_mmc_env_init(void)
 	run_command(cmd, 0);
 }
 
+#ifndef CONFIG_SPL_BUILD
 static pcb_rev_t get_pcb_revision(void)
 {
 	unsigned int pcb_vers = 0;
@@ -291,6 +295,23 @@ U_BOOT_CMD(
 	select_dt_from_module_version, CONFIG_SYS_MAXARGS, 1, do_select_dt_from_module_version,
 	"\n", "    - select devicetree from module version"
 );
+#endif
+
+/*
+ * We release the UART in the SPL hand-off, but don't release it due
+ * to the bug in the u-boot proper hand off, as there
+ * won't be serial output in Linux
+ */
+#ifdef CONFIG_SPL_BUILD
+void board_quiesce_devices(void)
+{
+	const char *power_on_devices[] = {
+		"dma_lpuart1",
+	};
+
+	imx8_power_off_pd_devices(power_on_devices, ARRAY_SIZE(power_on_devices));
+}
+#endif
 
 int board_mmc_get_env_dev(int devno)
 {
@@ -336,5 +357,9 @@ int board_late_init(void)
 #endif
 #endif /* CONFIG_IMX_LOAD_HDMI_FIMRWARE_RX || CONFIG_IMX_LOAD_HDMI_FIMRWARE_TX */
 
+#ifndef CONFIG_SPL_BUILD
+	select_dt_from_module_version();
+#endif
+
 	return 0;
 }
diff --git a/board/toradex/apalis-imx8/spl.c b/board/toradex/apalis-imx8/spl.c
new file mode 100644
index 00000000000..1d548df43a4
--- /dev/null
+++ b/board/toradex/apalis-imx8/spl.c
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018 NXP
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#include <common.h>
+#include <dm.h>
+#include <spl.h>
+#include <dm/uclass.h>
+#include <dm/device.h>
+#include <dm/uclass-internal.h>
+#include <dm/device-internal.h>
+#include <dm/lists.h>
+#include <bootm.h>
+#include <init.h>
+
+DECLARE_GLOBAL_DATA_PTR;
+
+void spl_board_init(void)
+{
+	struct udevice *dev;
+
+	uclass_find_first_device(UCLASS_MISC, &dev);
+
+	for (; dev; uclass_find_next_device(&dev)) {
+		if (device_probe(dev))
+			continue;
+	}
+
+	board_early_init_f();
+
+	timer_init();
+
+	preloader_console_init();
+
+	puts("Normal Boot\n");
+}
+
+void spl_board_prepare_for_boot(void)
+{
+	board_quiesce_devices();
+}
+
+#ifdef CONFIG_SPL_LOAD_FIT
+int board_fit_config_name_match(const char *name)
+{
+	/* Just empty function now - can't decide what to choose */
+	debug("%s: %s\n", __func__, name);
+
+	return 0;
+}
+#endif
+
+void board_init_f(ulong dummy)
+{
+	/* Clear the BSS. */
+	memset(__bss_start, 0, __bss_end - __bss_start);
+
+	arch_cpu_init();
+
+	board_init_r(NULL, 0);
+}
diff --git a/configs/apalis-imx8_defconfig b/configs/apalis-imx8_defconfig
index 876cccf2059..32e07aa1808 100644
--- a/configs/apalis-imx8_defconfig
+++ b/configs/apalis-imx8_defconfig
@@ -21,6 +21,7 @@ CONFIG_SYS_EXTRA_OPTIONS="IMX_CONFIG=board/toradex/apalis-imx8/apalis-imx8-imxim
 CONFIG_LOG=y
 # CONFIG_DISPLAY_BOARDINFO is not set
 CONFIG_BOARD_EARLY_INIT_F=y
+CONFIG_SYS_MMCSD_FS_BOOT_PARTITION=0
 CONFIG_CMD_CPU=y
 # CONFIG_BOOTM_NETBSD is not set
 CONFIG_CMD_ASKENV=y
diff --git a/include/configs/apalis-imx8.h b/include/configs/apalis-imx8.h
index 62ad6df0f79..746e1d8f427 100644
--- a/include/configs/apalis-imx8.h
+++ b/include/configs/apalis-imx8.h
@@ -9,6 +9,36 @@
 #include <asm/arch/imx-regs.h>
 #include <linux/sizes.h>
 
+#ifdef CONFIG_SPL_BUILD
+#define CONFIG_SPL_MAX_SIZE				(192 * 1024)
+#define CONFIG_SYS_MONITOR_LEN				(1024 * 1024)
+#ifndef CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_USE_SECTOR
+#define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_USE_SECTOR
+#define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR		0x1040 /* (flash.bin_offset + 2Mb)/sector_size */
+#endif
+
+/*
+ * 0x08081000 - 0x08180FFF is for m4_0 xip image,
+ * 0x08181000 - 0x008280FFF is for m4_1 xip image
+  * So 3rd container image may start from 0x8281000
+ */
+#define CONFIG_SYS_UBOOT_BASE 0x08281000
+
+#define CONFIG_SPL_LDSCRIPT		"arch/arm/cpu/armv8/u-boot-spl.lds"
+#define CONFIG_SPL_STACK		0x013fff0
+#define CONFIG_SPL_BSS_START_ADDR	0x00130000
+#define CONFIG_SPL_BSS_MAX_SIZE		0x1000	/* 4 KB */
+#define CONFIG_SYS_SPL_MALLOC_START	0x82200000
+#define CONFIG_SYS_SPL_MALLOC_SIZE	0x80000	/* 512 KB */
+#define CONFIG_SERIAL_LPUART_BASE	0x5a070000
+#define CONFIG_MALLOC_F_ADDR		0x00138000
+
+#define CONFIG_SPL_RAW_IMAGE_ARM_TRUSTED_FIRMWARE
+
+#define CONFIG_SPL_ABORT_ON_RAW_IMAGE
+
+#endif
+
 #define CONFIG_REMAKE_ELF
 
 #define CONFIG_DISPLAY_BOARDINFO_LATE
@@ -16,7 +46,6 @@
 #define CONFIG_SYS_FSL_ESDHC_ADDR	0
 #define USDHC1_BASE_ADDR		0x5b010000
 #define USDHC2_BASE_ADDR		0x5b020000
-#define CONFIG_SUPPORT_EMMC_BOOT	/* eMMC specific */
 
 #define CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG
 
@@ -46,8 +75,7 @@
 	func(MMC, mmc, 1) \
 	func(MMC, mmc, 2) \
 	func(MMC, mmc, 0) \
-	func(USB, usb, 0) \
-	func(DHCP, dhcp, na)
+	func(USB, usb, 0)
 #include <config_distro_bootcmd.h>
 
 #ifdef CONFIG_AHAB_BOOT
-- 
2.34.1

